tensorflow/contrib/tensorrt/convert/convert_graph.cc

```diff
@@ -692,16 +714,67 @@ tensorflow::Status ConvertAfterShapes(ConversionParams& params) {
           cuda_device_id = cuda_gpu_id.value();
         }
         tensorflow::GPUOptions gpuoptions;
-        // we need to us PM here since in python path there is no way to get
-        // to allocators
-        auto pm = tensorflow::ProcessState::singleton();
         // this should be instantiated by now
         auto dev_allocator = pm->GetGPUAllocator(gpuoptions, tf_gpu_id, 1);
         VLOG(1) << "Got an allocator for device tf_device=" << tf_gpu_id.value()
                 << " cuda device= " << cuda_device_id << " at "
                 << dev_allocator;
         alloc.reset(new TRTDeviceAllocator(dev_allocator));
       }
+    } else {
+      int found_device = 0;
+      bool try_gpu_ids = true;
+      auto checkDeviceId = [](int tfid) -> int {
+        tensorflow::TfGpuId tf_gpu_id(tfid);
+        CudaGpuId cuda_gpu_id;
+        Status s = GpuIdManager::TfToCudaGpuId(tf_gpu_id, &cuda_gpu_id);
+        if (s.ok()) {
+          VLOG(1) << "Found TF GPU " << tf_gpu_id.value() << " at cuda device "
+                  << cuda_gpu_id.value();
+          return cuda_gpu_id.value();
+        }
+        VLOG(2) << "TF GPU with id " << tfid << " do not exist " << s;
+        return -1;
+      };
+      // if device is set, try to find the device. Might be a problem for multi
+      // host case but TensorRT do not support multi host setups yet.
+      if (!engine.device.empty()) {
+        auto res = str_util::Split(engine.device, ":");
+        if (res.size() > 0) {
+          tensorflow::StringPiece s(res.back());
+          tensorflow::str_util::RemoveWhitespaceContext(&s);
+          uint64 dev_id = 0;
+          if (str_util::ConsumeLeadingDigits(&s, &dev_id)) {
+            found_device = dev_id;
+            cuda_device_id = checkDeviceId(found_device);
+            if (cuda_device_id >= 0) try_gpu_ids = false;
+          }
+        }
+      }
+      if (try_gpu_ids) {
+        while (found_device < 100) {
+          cuda_device_id = checkDeviceId(found_device);
+          if (cuda_device_id >= 0) {
+            break;
+          }
+          found_device++;
+        }
+      }
+      if (found_device == 100) {
+        LOG(ERROR) << " Can't find a GPU device to work with. Please "
+                      "instantiate a session to initialize devices";
+        return tensorflow::errors::NotFound(
+            "Can't find a GPU device to work with");
+      }
+      LOG(WARNING)
+          << "Can't determine the device constructing an allocator at device "
+          << found_device;
+      tensorflow::GPUOptions gpuoptions;
```

Comment:
Please don't create the allocator here. Remember that allocator is process-wide resource, so the allocator created here will conflict with the ones created by the user explicitly via tf.Session() or any higher level API that will create the session. For now, let's return an error at line 763 or use the TRTCudaAllocator instead, and add a TODO(aaroey), we'll figure out how to find the allocator in this case.

Reply:
Are you sure, If user created an allocator on the found device, pm->GetGPUAllocator() will return it and won't construct an allocator. Only problem could be that the if this is instantiated first. It will construct default allocator with allow_growth option enabled. Do we still want that?

Reply:
Yes, if user created an allocator before this will return that one instead of creating a new one. Also, GetGPUAllocator accepts a TfGpuId whose value is **NOT** the device id, so the allocator it returns could be associated with any device, not necessary the one you want here.

Reply:
And, with the cluster problem in OptimizeGraph() fixed, I think we don't need this path any more as we can always get the device.

Reply:
>Also, GetGPUAllocator accepts a TfGpuId whose value is NOT the device id, 
I doubt it since device is choosen by querying the GpuIdManager::TfToCudaGpuId(tf_gpu_id, &cuda_gpu_id); and process manager uses the same function to map between tf and cuda gpu ids.

>And, with the cluster problem in OptimizeGraph() fixed, I think we don't need this path any more as we can always get the device.
Is this available right now?

Reply:
The cluster fix should be available with next sync.

Reply:
Ok. We can update this with a new PR after it is synced then